pkg/3scale/amp/auto-generated-templates/amp/amp-eval.yml

```diff
@@ -3970,7 +3960,7 @@ objects:
         - apicast-staging
         from:
           kind: ImageStreamTag
-          name: amp-apicast:latest
+          name: amp-apicast:${AMP_RELEASE}
```

Comment:
Hi,

I think we still need to keep the ImageChange trigger.

A ConfigChange trigger defined in a DeploymentConfig (https://docs.openshift.com/container-platform/3.11/dev_guide/deployments/basic_deployment_operations.html#config-change-trigger) results in a new replication controller whenever changes are detected in the PodTemplate of the deployment configuration.

A ImageChange trigger defined in a DeploymentConfig (https://docs.openshift.com/container-platform/3.11/dev_guide/deployments/basic_deployment_operations.html#image-change-trigger) results in a new replication controller whenever the content of an image stream tag changes (when a new version of the image is pushed).

Between 3scale releases (2.7, 2.8, ...) no ImageChange trigger would be needed in case we decide to directly set the image in the 'image' field of the DeploymentConfig and not in the ImageChangeTrigger field.
However, take into account the patch releases, where the AMP_RELEASE does NOT change. The only way to incorporate those changes then is through the refreshment of the ImageStream, and the way to update the DeploymentConfig automatically when an ImageStream changes is by using an ImageChange trigger.

Reply:
I've made some additional tests:

If I change a configuration field of the PodTemplateSection of the DeploymentConfig, a redeployment is performed due to the DeploymentConfig Config change trigger. The "generation" metadata field of the DC is incremented. Here the status can be seen:
```
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-02-25T10:42:19Z"
    lastUpdateTime: "2020-02-25T10:42:19Z"
    message: Deployment config has minimum availability.
    status: "True"
    type: Available
  - lastTransitionTime: "2020-02-25T11:13:06Z"
    lastUpdateTime: "2020-02-25T11:13:08Z"
    message: replication controller "backend-worker-7" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
  details:
    causes:
    - type: ConfigChange
    message: config change
  latestVersion: 7
  observedGeneration: 9
  readyReplicas: 1
  replicas: 1
  unavailableReplicas: 0
  updatedReplicas: 1
```

If I change the imageChangeTrigger referenced image in the DeploymentConfig, a redeployment is performed due to the DeploymentConfig Image change trigger. The "generation" metadata field of the DC is incremented. Here the status can be seen:
```
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-02-25T10:42:19Z"
    lastUpdateTime: "2020-02-25T10:42:19Z"
    message: Deployment config has minimum availability.
    status: "True"
    type: Available
  - lastTransitionTime: "2020-02-25T11:07:06Z"
    lastUpdateTime: "2020-02-25T11:07:09Z"
    message: replication controller "backend-worker-6" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
  details:
    causes:
    - imageTrigger:
        from:
          kind: DockerImage
          name: quay.io/miguelsorianod/apisonator-test@sha256:83160cea70919e5ec71fa41277a580722683d9275c3b61382cca3c63eafeae7e
      type: ImageChange
    message: image change
  latestVersion: 6
  observedGeneration: 8
  readyReplicas: 1
  replicas: 1
  unavailableReplicas: 0
  updatedReplicas: 1
```

When an edit is performed on the DC modifying a PodTemplateSpec field and  the referenced image in the ImageChangeTrigger, both being modified at the same time (with oc edit) it has been observed that the generation field has been incremented by two, but I've only been able to see the redeployment of one pod (for example if the original pod was backend-worker-9-* the new one that I observe is backend-worker-10). Same happens with the associated ReplicationController. In the events of the DC I also see just one event of DeploymentCreated that indicates that a new replicationcontroller has been created. Additionally in the status section of the DC I see that in the status field it shows an ImageChange has happened:
```
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-02-25T10:42:19Z"
    lastUpdateTime: "2020-02-25T10:42:19Z"
    message: Deployment config has minimum availability.
    status: "True"
    type: Available
  - lastTransitionTime: "2020-02-25T11:35:37Z"
    lastUpdateTime: "2020-02-25T11:35:40Z"
    message: replication controller "backend-worker-10" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
  details:
    causes:
    - imageTrigger:
        from:
          kind: DockerImage
          name: quay.io/miguelsorianod/apisonator-test@sha256:83160cea70919e5ec71fa41277a580722683d9275c3b61382cca3c63eafeae7e
      type: ImageChange
    message: image change
  latestVersion: 10
  observedGeneration: 14
  readyReplicas: 1
  replicas: 1
  unavailableReplicas: 0
  updatedReplicas: 1
```

It does not show that a ConfigChange has happened, but the pod has the change applied to the PodTemplateSpec too.